contracts/contracts/lib/LibGateway.sol

```diff
@@ -547,4 +559,86 @@ library LibGateway {
             revert MaxMsgsPerBatchExceeded();
         }
     }
+
+    /// @notice Validates a cross message before committing it.
+    function validateCrossMessage(IpcEnvelope memory envelope) internal view returns (CrossMessageValidationOutcome) {
+        GatewayActorStorage storage s = LibGatewayActorStorage.appStorage();
+        SubnetID memory toSubnetId = envelope.to.subnetId;
+
+        if (toSubnetId.isEmpty()) {
+            return CrossMessageValidationOutcome.InvalidDstSubnet;
+        }
+
+        // We cannot send a cross message to the same subnet.
+        if (toSubnetId.equals(s.networkName)) {
+            return CrossMessageValidationOutcome.CannotSendToItself;
+        }
+
+        // Lowest common ancestor subnet
+        bool isLCA = toSubnetId.commonParent(envelope.from.subnetId).equals(s.networkName);
+        IPCMsgType applyType = envelope.applyType(s.networkName);
+        
+        // If the directionality is top-down, or if we're inverting the direction
+        // else we need to check if the common parent exists.
+        if (applyType == IPCMsgType.TopDown || isLCA) {
+            (bool foundSubnet,) = LibGateway.getSubnet(toSubnetId.down(s.networkName));
+            if (!foundSubnet) {
+                return CrossMessageValidationOutcome.InvalidDstSubnet;
+            }
+        } else {
+            SubnetID memory commonParent = toSubnetId.commonParent(s.networkName);
+            if (commonParent.isEmpty()) {
+                return CrossMessageValidationOutcome.CommonParentNotExist;
+            }
+        }
+
+        return CrossMessageValidationOutcome.Valid;
+    }
+    
+     /**
+     * @dev Propagates all the populated cross-net messages from the postbox.
+     */
+    function propagateAllPostboxMessages() internal {
+        GatewayActorStorage storage s = LibGatewayActorStorage.appStorage();
+
+        uint256 keysLength = s.postboxKeys.length();
+        bytes32[] memory ids = new bytes32[](keysLength);
+        for (uint256 i = 0; i < keysLength; ) {
+            bytes32 msgCid = s.postboxKeys.at(i);
+            ids[i] = msgCid;
+            LibGateway.propagatePostboxMessage(msgCid);
+
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+     /**
+     * @dev Propagates the populated cross-net message for the given `msgCid`.
+     * @param msgCid - the cid of the cross-net message
+     */
+    function propagatePostboxMessage(bytes32 msgCid) internal {
+        GatewayActorStorage storage s = LibGatewayActorStorage.appStorage();
+        IpcEnvelope storage crossMsg = s.postbox[msgCid];
+
+        if (crossMsg.isEmpty()) {
```

Comment:
will this ever happen？This should be checked before commitment?

Reply:
I think the only reason I kept this here is because Solidity doesn’t have any other way to check if a record in the mapping exists, right? Correct me if I’m wrong. Although, I guess we can trust that the keys we iterate through in postboxKeys are correctly in sync. So, it should be fine to remove it?